### ISSUE TITLE
Display of large vectors kills editor

### ISSUE THREAD

**Issue by nilshg** [JunoLab/atom-julia-client#12]
Are other people seing this?

`rand(1000)` takes close to no time, `rand(10000)` takes substantially longer, while `rand(100000)` kills the editor for me (i.e. I get a couple of "Editor is not reacting, do you want to wait?" messages, but it doesn't recover). 

When I evaluate e.g.
````
rand(100000)
1
````
so that the output being displayed is just `1`, rather than the large vector, evaluation again happens instantaneously, suggesting that the large vector is causing the problem.

**Comment by malmaud** [comment]
I just came here to report this.

**Comment by Varanas** [comment]
That behavior makes sense -- we should probably truncate the displayed output after some sensible threshold (just like Julia's REPL). That shouldn't be much of a problem, since no one would really want to scroll through 100,000 elements, right? :-)

**Comment by one-more-minute** [comment]
Yup, the problem here is that Julia tries to display *every* element of the vector which crashes Atom (possibly just because it takes so long to transfer the result over TCP). In future I'd really like to have a tree-like display for large results, e.g. something like:
```
v Vector Int, 100
  v 1 - 10
    1
    4
    6
    ...
  > 11 - 20
  > 21 – 30
  ...
```
However, that needs lazy-loading to work really well. A reasonable stopgap would be to just cut out the middle elements, as Juno/LT and the repl do currently.

**Comment by nilshg** [comment]
I was actually going to do just this (the stopgap measure), but then I realised that my CoffeeScript knowledge is so woefully inadequate that I couldn't even locate the part of the code I'd have to change...

**Comment by one-more-minute** [comment]
Fortunately, you wouldn't have to touch CoffeeScript at all – the tree views are all created in Julia, [here](https://github.com/JunoLab/Atom.jl/blob/fcafda674675ca61120d7293ee10cb5c5c59e908/src/display/objects.jl#L42-L46), and then sent to the frontend to be rendered. It's as simple as cutting out some of the children if there are too many.

In general, if you're interested in patching anything but not sure where to start, feel free to let me know and I'll try to give pointers. I should probably write up some more devdocs on how display works so that it's a bit easier to get into.

**Comment by nilshg** [comment]
Okay, I don't really understand what's going on here, which is why I won't submit a PR (where is `render` defined? And how does it differ from the macro `@render`? Where is the `Inline` type defined?), but a working version of what I'd like to see would be:
````
@render i::Inline xs::Vector begin
  if length(xs) <= 25
    Tree(span(strong("Vector"),
            fade(" $(eltype(xs)), $(length(xs))")),
       [render(i, x, options = options) for x in xs])
  else
    Tree(span(strong("Vector"),
            fade(" $(eltype(xs)), $(length(xs))")),
       [[render(i, x, options = options) for x in xs[1:10]];
        "...";
        [render(i, x, options = options) for x in xs[end-9:end]]])
  end
end
````

**Comment by one-more-minute** [comment]
I'd be happy to take that as a PR, perhaps with just a little refactoring to avoid the repetition.

I'll also write up a bit more about how display is working and point you to it when I get the chance.

**Comment by nilshg** [comment]
Okay, not sure my refactoring is what you were looking for, but I've created a PR.

**Comment by one-more-minute** [comment]
The original issue is fixed so I'm re-purposing this for the tree views thing I mentioned above.
